Add SCIM service discovery endpoints

robfusion opened the pull request:
Add documentation for the 3 service discovery endpoints and fixed a couple of messed up links I found along the way.

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
+[.endpoint]
+.URI
+--
+[method]#GET# [uri]#/api/scim/resources/v2/ResourceTypes#
```
```suggestion
[method]#GET# [uri]#/api/scim/resource/v2/ResourceTypes#
```

mooreds replied:
Need unauthenticated icon here.

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
+[.endpoint]
+.URI
+--
+[method]#GET# [uri]#/api/scim/resources/v2/Schemas/`\{schemaId\}`#
```
```suggestion
[method]#GET# [uri]#/api/scim/resource/v2/Schemas`\{schemaId\}`#
```

mooreds replied:
need unauthenticated icon here.

mooreds commented on site/docs/v1/tech/apis/scim/_scim-response-codes.adoc:
```diff
@@ -14,21 +14,25 @@ ifdef::delete[]
+endif::[]
 
-ifndef::create[]
+ifndef::no_missing[]
```
`never_missing` seems more typical.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-response-codes.adoc:
```diff
@@ -14,21 +14,25 @@ ifdef::delete[]
 |500
 |There was an internal error. A stack trace is provided and logged in the FusionAuth log files. The response will be empty.
 
+ifndef::no_search[]
```
`never_search_error` seems to be more standard

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-response-codes.adoc:
```diff
@@ -14,21 +14,25 @@ ifdef::delete[]
 |The search index is not available or encountered an exception so the request cannot be completed. The response will contain a JSON body.
+endif::[]
 
 ifndef::no_webhook_event[]
```
The inverse seems to be more common: `webhook_event` as illustrated here: `site/docs/v1/tech/apis/_standard-post-response-codes.adoc`

robfusion replied:
In the case of SCIM reponse codes. webhooks are much more common because almost all the endpoints are CRUD endpoints for Users or Group Members.

mooreds replied:
Sounds good.

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
 
-This API is used to retrieve information about the configuration of FusionAuth SCIM Service Provider per the SCIM specification.
+:unauthenticated:
+:no_missing:
```
I was able to get a 404 status code for schemas, if not for the others.

curl -vvv -X GET https://sandbox.fusionauth.io/api/scim/resource/v2/Schemas/abc

robfusion replied:
ah, good point. I was thinking of the root endpoint not getting a 404

robfusion replied:
fixed

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
+[.endpoint]
+.URI
+--
 [method]#GET# [uri]#/api/scim/resources/v2/ServiceProviderConfig#
```
```suggestion
[method]#GET# [uri]#/api/scim/resource/v2/ServiceProviderConfig#
```

mooreds replied:
Need unauthenticated icon

mooreds commented on site/docs/src/json/scim/scim-resourcetypes-response.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "totalResults": 2,
```
These should be alphabetized.

robfusion replied:
done

mooreds commented on site/docs/src/json/scim/scim-schemas-response.json:
```diff
@@ -0,0 +1,208 @@
+{
+  "itemsPerPage": 4,
```
Should be alphabetized. jq does it well.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-response-codes.adoc:
```diff
@@ -14,21 +14,25 @@ ifdef::delete[]
-|The request was invalid and/or malformed. The response will contain an link:/docs/v1/tech/apis/scim[SCIM Errors] JSON Object with the specific errors. This status will also be returned if a paid FusionAuth license is required and is not present.
+|The request was invalid and/or malformed. The response will contain a link:/docs/v1/tech/apis/scim[SCIM Error] JSON Object with the specific errors. This status will also be returned if a paid FusionAuth license is required and is not present.
 
+ifndef::unauthenticated[]
```
`no_authorization` seems to be what is defined here: site/docs/v1/tech/apis/_standard-get-response-codes.adoc 

Please use that.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
+[.endpoint]
+.URI
+--
 [method]#GET# [uri]#/api/scim/resources/v2/Schemas#
```
```suggestion
[method]#GET# [uri]#/api/scim/resource/v2/Schemas#
```

mooreds commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
+
+[.api]
+[field]#schemaId# [type]#[String]# [optional]#Optional#::
+The unique Schema Id.
```
```suggestion
The unique Schema Id, such as `urn:ietf:params:scim:schemas:core:2.0:User`.
```

mooreds commented on site/docs/v1/tech/apis/scim/_scim-schemas-response-body.adoc:
```diff
@@ -0,0 +1,18 @@
+[NOTE.note]
```
Break note out to include file.

robfusion replied:
done

mooreds commented on site/docs/v1/tech/apis/scim/_scim-resourcetypes-response-body.adoc:
```diff
@@ -0,0 +1,18 @@
+[NOTE.note]
```
Break note out to include file.

robfusion replied:
done

mooreds reviewed (changes requested):
Some general comments:

need to add ServiceProvider page to left hand nav

The icon here http://localhost:4000/docs/v1/tech/apis/authentication#client-credentials is different than icon here: http://localhost:4000/docs/v1/tech/apis/scim/SCIMGroup#create-a-group

A general question: how do you protect SCIM if there is no auth required for the schema endpoints? Network layer? IP address locking?

robfusion commented:
> 

The discovery endpoints aren't meant to be authenticated. They only show the configuration of the service provider. Similar to getting /.well-known

robfusion commented on site/docs/v1/tech/apis/scim/SCIMServiceProvider.adoc:
```diff
@@ -4,12 +4,94 @@ title: SCIM Service Provider API Overview
 --
+
+Retrieve a Schema by Id
+[.endpoint]
```
They are unautheticated.

robfusion replied:
done

mooreds commented:
@robfusion re: https://github.com/FusionAuth/fusionauth-site/pull/1339#issuecomment-1110255746 sure, but even for unauthenticated endpoints we have an icon. See https://fusionauth.io/docs/v1/tech/apis/users#change-a-users-password for example. That unlocked icon points to https://fusionauth.io/docs/v1/tech/apis/users#change-a-users-password

robfusion commented on site/docs/v1/tech/apis/scim/SCIMEnterpriseUser.adoc:
```diff
@@ -34,9 +34,9 @@ include::/docs/v1/tech/apis/scim/_scim-enterpriseuser-request-body.adoc[]
 The response for this API contains the User that was just created in SCIM schema format.
 
-:create:
+:no_missing:
```
done

robfusion commented on site/docs/v1/tech/apis/scim/_scim-resourcetypes-response-body.adoc:
```diff
@@ -0,0 +1,18 @@
+Because the SCIM specification allows for customization of the schemas using extensions, there is no way to accurately document all the JSON structure possibilities.
+====
+
+The following is the default response body. However, you can customize the schemas portion of the response by modifying the schemas setting link:/docs/v1/tech/core-concepts/tenants#scim[here].
```
fixed

robfusion commented on site/docs/v1/tech/apis/scim/_scim-resourcetypes-response-body.adoc:
```diff
@@ -0,0 +1,18 @@
+
+The following is the default response body. However, you can customize the schemas portion of the response by modifying the schemas setting link:/docs/v1/tech/core-concepts/tenants#scim[here].
+
+This is taken from the SCIM RFC describing a SCIM ResourceTypes schema. For the full specification you can find the RFC link:https://datatracker.ietf.org/doc/html/rfc7643#section-6[here].
```
fixed

robfusion commented on site/docs/v1/tech/apis/scim/_scim-serviceproviderconfig-response-body.adoc:
```diff
@@ -0,0 +1,16 @@
+[NOTE.note]
+====
```
done

mooreds commented:
A few issues:

* the icons are still wrong for http://localhost:4000/docs/v1/tech/apis/scim/SCIMGroup vs http://localhost:4000/docs/v1/tech/apis/authentication#client-credentials (is that a separate issue?)

* there is no nav entry for the SCIM Service Provider API Overview

* GET /api/scim/resource/v2/Schemas`{schemaId}` renders wrong on http://localhost:4000/docs/v1/tech/apis/scim/SCIMServiceProvider (I think you need a `/` before `schemaId` and for some reason the backticks show up.

* would you recommend these endpoints be unsecured in a prod deployment? (Is that a separate issue?)

robfusion commented:
> 

Sorry, missed the nav change and the wrong icon on client creds. I'll get those. The unsecured endpoints is probably a too big a conversation for the comments. We should bring in Daniel for that too.

robfusion commented:
> @robfusion re: [#1339 (comment)](https://github.com/FusionAuth/fusionauth-site/pull/1339#issuecomment-1110255746) sure, but even for unauthenticated endpoints we have an icon. See https://fusionauth.io/docs/v1/tech/apis/users#change-a-users-password for example. That unlocked icon points to https://fusionauth.io/docs/v1/tech/apis/users#change-a-users-password

I thought I got the unauthenticated icon on all of them. Did I miss some?

robfusion commented:
> A few issues:
> 
> * the icons are still wrong for http://localhost:4000/docs/v1/tech/apis/scim/SCIMGroup vs http://localhost:4000/docs/v1/tech/apis/authentication#client-credentials (is that a separate issue?)
> * there is no nav entry for the SCIM Service Provider API Overview
> * GET /api/scim/resource/v2/Schemas`{schemaId}` renders wrong on http://localhost:4000/docs/v1/tech/apis/scim/SCIMServiceProvider (I think you need a `/` before `schemaId` and for some reason the backticks show up.
> * would you recommend these endpoints be unsecured in a prod deployment? (Is that a separate issue?)

* added to the nav under APIs -> SCIM
* I can't figure out why the icon shows up as a different color. The code is the same in both places. Might need some FTL expertise on that one.
* Fixed the wrong render on the request param.

mooreds reviewed (approved):
LGTM!
